pkg/app/piped/platformprovider/ecs/diff.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecs
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/pipe-cd/pipecd/pkg/diff"
+)
+
+const (
+	diffCommand = "diff"
+)
+
+type DiffResult struct {
+	Diff *diff.Result
+	Old  ECSManifests
+	New  ECSManifests
+}
+
+func (d *DiffResult) NoChange() bool {
+	return len(d.Diff.Nodes()) == 0
+}
+
+func Diff(old, new ECSManifests, opts ...diff.Option) (*DiffResult, error) {
+	d, err := diff.DiffStructureds(old, new, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if !d.HasDiff() {
+		return &DiffResult{Diff: d}, nil
+	}
+
+	ret := &DiffResult{
+		Old:  old,
+		New:  new,
+		Diff: d,
+	}
+	return ret, nil
+}
+
+type DiffRenderOptions struct {
+	// If true, use "diff" command to render.
+	UseDiffCommand bool
+}
+
+func (d *DiffResult) Render(opt DiffRenderOptions) string {
+	var b strings.Builder
+	opts := []diff.RenderOption{
+		diff.WithLeftPadding(1),
+	}
+	renderer := diff.NewRenderer(opts...)
+	if !opt.UseDiffCommand {
+		b.WriteString(renderer.Render(d.Diff.Nodes()))
+	} else {
+		d, err := diffByCommand(diffCommand, d.Old, d.New)
+		if err != nil {
+			b.WriteString(fmt.Sprintf("An error occurred while rendering diff (%v)", err))
+		} else {
+			b.Write(d)
+		}
+	}
+	b.WriteString("\n")
+
+	return b.String()
+}
+
+func diffByCommand(command string, old, new ECSManifests) ([]byte, error) {
+	taskDiff, err := diff.DiffByCommand(command, old.TaskDefinition, new.TaskDefinition)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceDiff, err := diff.DiffByCommand(command, old.ServiceDefinition, new.ServiceDefinition)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes.Join([][]byte{
+		[]byte("# 1. ServiceDefinition"),
+		serviceDiff,
+		[]byte("\n# 2. TaskDefinition"),
+		taskDiff,
+	}, []byte("\n")), nil
+}
```

Comment:
thanks, I fixed it. [0cb5d29](https://github.com/pipe-cd/pipecd/pull/4881/commits/0cb5d29436de848192c1ae45b3a6bc9ed2bd33d1)

`renderByCommand` is better especially because this func contains the following code.

```
return bytes.Join([][]byte{
		[]byte("# 1. ServiceDefinition"),
		serviceDiff,
		[]byte("\n# 2. TaskDefinition"),
		taskDiff,
	}, []byte("\n")), nil
```